resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala

```diff
@@ -144,6 +144,16 @@ private[spark] class Client(
       if (waitForAppCompletion) {
         logInfo(s"Waiting for application $appName to finish...")
         loggingPodStatusWatcher.awaitCompletion()
+        try {
+          // When the submission client sees that the application finished, it deletes all
+          // Kubernetes resources the application depends on instead of relying on Kubernetes
+          // garbage collection to kick in on the resources, which only happens when the driver
+          // pod gets explicitly deleted.
+          val otherKubernetesResources = currentDriverSpec.otherKubernetesResources
```

Comment:
One other caveat is that we lose some state that might be useful for debugging the driver. For example, we lose the ConfigMap object that is required to set up the init-container, which is useful for knowing what the properties resolved to if the init-container fails. I'm not sure how problematic this will be in practice, but it is something to consider.

Reply:
The ConfigMap is purely an implementation detail and really shouldn't be exposed to the users. For the purpose of debugging, I think it's better logging the ConfigMap contents instead of leaving the ConfigMap around after an application finished.

Reply:
ConfigMap aside, I think the notion of keeping pod for debugging make sense, I thought we had a config on that (it might be a different area/code path)